geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java

```diff
@@ -749,7 +751,8 @@ protected boolean hasExpired(long msTillHeadTombstoneExpires) {
         testHook_forceExpirationCount--;
         return true;
       }
-      return msTillHeadTombstoneExpires <= 0;
+      // In case
+      return msTillHeadTombstoneExpires <= 0 || msTillHeadTombstoneExpires > EXPIRY_TIME;
```

Comment:
I maybe missing the math but long msTillHeadTombstoneExpires = oldest.getVersionTimeStamp() + EXPIRY_TIME - now;
the only time msTillHeadTombstoneExpires is greater than EXPIRY_TIME will be when oldest.timeStamp - now is a postive number. Can this occur?

Reply:
That is the problem we are seeing. The customer is having timestamps far into the future. That is why we are putting this change in... So once we subtract now from oldest.getVersionTimeStamp() the number is far greater than EXPIRY_TIME. This is intended to catch that condition...

Reply:
It seems like this fix is just masking the real issue, which is that timestamps shouldn't be far in the future in the first place. Do we have any idea why/how that might be happening?

Reply:
For such a workaround, we should put detail comments here

Reply:
Yup, I lost the comments some how, so I am putting them in now. I agree this abates the symptom. I don't prefer this approach, but at the same time, this will help people with their problem that should not be happening. We were suspicious this had to do with clock time adjustments, but we are still theorizing. This will keep their system from locking up though for a condition we can safely handle in this part of the system with sanity checking. Otherwise intervention is required when this problem arises. Also to be blunt, the current approach while fast is lacking in fault tolerance.